File: NodeLink/138. 复制带随机指针的链表.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
 var copyRandomList = function(head) {
    // 解题思路：因为节点都有一个random指针，所以我们需要使用在每一个节点的后面插入一个节点同时将random指针指向插入前的节点所指向的节点，在循环将每一个插入的节点额random指针调整为random.next节点，最后在将插入的节点拆分出来组成新的链表则完成复制
    if(head == null) return head;
    let p = head
    let q;
    // 
    while(p!=null){
        q = new Node(p.val,null,null)
        q.random = p.random;
        q.next = p.next;
        p.next = q;
        p = q.next
    }
    q = head.next;
    while(q !== null){
        if(q.random !== null) q.random = q.random.next;
        // 首先将指针向后移动一位如果不存在则结束，如果存在则再往后移动一位
        (q = q.next) && (q = q.next)
        
    }

   let new_list = head.next
    p = head
    while(p !== null){
        q = p.next;
        p.next = q.next;
        if(p.next !== null) q.next = p.next.next;  
        p = p.next;
    }
    return new_list

};4 

File: Python/队列/138. 复制带随机指针的链表.py
# -*- codeing = utf-8 -*-
# @Time : 2023-9-8 10:35
# @Author : 刘永奇
# @File : 138. 复制带随机指针的链表.py
# @Software : PyCharm
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""


class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head: return None
        q = head
        while q:
            p = Node(q.val)
            p.next = q.next
            p.random = q.random
            q.next = p
            q = p.next
        p = head.next
        while p:
            if p.random: p.random = p.random.next
            p = p.next
            if p: p = p.next
        new_head = head.next
        p = head
        while p:
            q = p.next
            p.next = q.next
            if q.next: q.next = p.next.next
            p = p.next
        return new_head

